### PR TITLE
fix: guard trunk sampler against NaN logits at long context

### PR DESCRIPTION
**Pull request by daniel-farina**
Tracking issue: youssofal/MTPLX#49.

## Diagnosis

The trunk logit tensor can contain NaN/Inf at long context (typically >25K tokens with sustained profile) due to fp16 cancellation in dense SDPA fallback paths (`cache_state.py:_large_q_split_sdpa_fallback` or `attention_split.py:264`'s SDPA over ~25K-row K). This propagates through the sampler:

```
mx.logsumexp(flat, axis=-1)       -> NaN
mx.exp(top_vals - log_total)      -> NaN probabilities
probs.sum() <= 0                  -> rescue check SKIPPED (IEEE-754: NaN<=0 is False)
SparseDistribution(probs=[NaN])   -> raises at sampling.py:39
```

Result: `ValueError/500: SparseDistribution probabilities must have positive mass`. When only some logit positions are NaN, sampling instead picks an off-grammar token and produces a truncated `<tool_call>` block (`422 malformed tool_call`). Same root cause, two failure modes.

## Fix (sampler-boundary mitigation)

- Sanitize non-finite logits with `mx.where(mx.isfinite, flat, -1e30)` before `argpartition`/`logsumexp` in all three `fast_sampling.py` entry points.
- Replace `<= 0` rescue checks with `not (>0)` so NaN totals trip rescue (IEEE-754 fix).
- Defensive per-row one-hot collapse if residual NaN survives upstream sanitize.
- Mirror the fix in the numpy `softmax` path at `sampling.py:63`.
- Env gate `MTPLX_TRUNK_LOGITS_NAN_GUARD=0` to disable for debugging; `MTPLX_TRUNK_LOGITS_NAN_GUARD_QUIET=1` to silence per-fire log line.

## Empirical validation

Same bricked 27K coding-agent session before/after:

| Probe | Before fix | After fix |
|---|---|---|
| One-shot at 26K | 500 `SparseDistribution` OR 15min stream timeout | 40s, `finish_reason=stop`, ctok=80, real answer |
| Tool-using turn at 26K | 422 `malformed tool_call` on round 1 | 9 valid tool-calling rounds, 422 on round 9, recovered |
| Push to 29K | unreachable | reachable, productivity degrades but no crashes |

The 422 failure mode still surfaces occasionally (the degenerate-but-not-fully-NaN form), but at lower frequency, and is recoverable client-side. The hard 500 NaN is gone.

## Scope

This is intentionally a sampler-boundary mitigation, not the upstream attention fix. The fp16 cancellation in dense SDPA at long context is a deeper bug that wants instrumented forward-pass diagnostics; #49 tracks it as a follow-up.

## Tests

13 sampling/fast-sampling tests pass, including 6 new regression tests covering single-NaN, all-NaN, batched per-row NaN, and full-row NaN cases.

Stacked on `daniel/dev-stack` which also rolls up #44 + #46 + #47 + #48.

### CODE REVIEW

**Comment by daniel-farina** [comment]
Quick status while waiting for review:

Empirically the sampler-boundary NaN guard in this PR is doing what it claims — at >25K context the trunk distribution is degenerate often enough that without the guard we get hard `SparseDistribution: positive mass` 500s, and with it we get either a recovered token or a clear malformed-tool_call which clients can drop and retry. So it's a useful mitigation.

That said, we're now investigating the upstream attention numerical issue that's causing the degenerate trunk logits in the first place — this PR doesn't fix that, it just contains the symptom. If our investigation finds a clean upstream fix, we'll prefer that and may end up closing this PR in favor of an attention-side patch. If it turns out fp16 cancellation in the long-context dense SDPA path is intrinsic and not fixable cheaply, then this PR's mitigation is the right shape and we'll ask for review.

Marking this as WIP-status-comment, no code changes; will reply again with the diagnosis.

**Comment by daniel-farina** [comment]
Closing this PR.

Spent the past hour empirically analyzing 1557 generations across the dev-stack server. Result: **zero NaN-guard fires recorded**. The sampler-boundary mitigation in this PR is correct in spirit but never actually triggers in practice — the supposed long-context cliff isn't a NaN, it's just O(N²) prefill latency on the 16 global-attention layers of qwen3.6-27b crossing client-side timeouts.

What we thought was a numerical-stability bug:
- malformed unclosed `<tool_call>` blocks → degenerate sampling, 1.5% rate, fix is sampler constraint not NaN guard
- truncated streams at >25K → just slow prefill (43K-token prompts complete in ~80s)
- `mtplx_stream_silence` events → fire during prefill, not mid-stream stall (`server/openai.py:6228`)

PR #44 (the bench abort relax) was the right fix because the abort genuinely fired. PR #46 (env override) was the right fix because the env var was silently dropped. This PR was solving a problem that doesn't exist on this setup.

Net: closing as superseded by 'don't panic, just wait for prefill'. The real long-context issue is throughput, not numerical, and the fix is client-side timeout tolerance, not server-side mitigation. If a future setup actually produces NaN in attention, the diagnosis stack here will help find it quickly.
